packages/drivers/odsp-driver/src/odspSummaryUploadManager.ts

```diff
@@ -0,0 +1,412 @@
+
+    constructor(
+        private readonly snapshotUrl: string,
+        private readonly getStorageToken: (options: TokenFetchOptions, name?: string) => Promise<string | null>,
+        private readonly logger: ITelemetryLogger,
+        private readonly epochTracker: EpochTracker,
+        private readonly blobCache: Map<string, IBlob | ArrayBuffer> = new Map(),
+    ) {
+    }
+
+    /**
+     * Builts the caches which will be used for blob deduping.
+     * @param snapshotTree - snapshot tree from which the dedup caches are built.
+     */
+    public async buildCachesForDedup(snapshotTree: api.ISnapshotTree) {
+        const prefixedSnapshotTree = this.addAppPrefixToSnapshotTree(snapshotTree);
+        await this.buildCachesForDedupCore(prefixedSnapshotTree);
+    }
+
+    /**
+     * Builts the caches which will be used for blob deduping.
+     * @param snapshotTree - snapshot tree from which the dedup caches are built.
+     * @param path - path of the current node evaluated.
+     */
+    private async buildCachesForDedupCore(snapshotTree: api.ISnapshotTree, path: string = ""): Promise<api.ISummaryTree> {
+        assert(Object.keys(snapshotTree.commits).length === 0, "There should not be commit tree entries in snapshot");
+
+        const summaryTree: api.ISummaryTree = {
+            type: api.SummaryType.Tree,
+            tree: {},
+        };
+        for (const [key, value] of Object.entries(snapshotTree.blobs)) {
+            const blobValue = this.blobCache.get(value);
+            assert(blobValue !== undefined, "Blob should exists");
+            const hash = await hashFile(
+                blobValue instanceof ArrayBuffer ?
+                    IsoBuffer.from(blobValue) :
+                        IsoBuffer.from(blobValue.content, blobValue.encoding ?? "utf-8"),
+            );
+            // We are setting the content as empty because we won't use it anywhere.
+            // Instead we will use the hash of the blob from pathToBlobSha cache.
+            summaryTree.tree[key] = {
+                type: api.SummaryType.Blob,
+                content: "",
+            };
+            // fullBlobPath does not start with "/"
+            const fullBlobPath = path === "" ? key : `${path}/${key}`;
+            this.blobTreeDedupCaches.blobShaToPath.set(hash, fullBlobPath);
+            this.blobTreeDedupCaches.pathToBlobSha.set(fullBlobPath, hash);
+        }
+
+        for (const [key, tree] of Object.entries(snapshotTree.trees)) {
+            // fullTreePath does not start with "/"
+            const fullTreePath = path === "" ? key : `${path}/${key}`;
+            const subtree = await this.buildCachesForDedupCore(tree, fullTreePath);
+            this.blobTreeDedupCaches.treesPathToTree.set(fullTreePath, subtree);
+            summaryTree.tree[key] = subtree;
+        }
+        return summaryTree;
+    }
+
+    private addAppPrefixToSnapshotTree(snapshotTree: api.ISnapshotTree): api.ISnapshotTree {
```

Comment:
Can you please add comment on the purpose of this function / what it does?

Reply:
done